### ISSUE TITLE
Bug: running `prisma generate` sometimes gets stuck and have to juggle between different ways the prisma `provider` script is written

### ISSUE THREAD

**Issue by capaj** [YassinEldeeb/prisma-tgql-types-gen#58]
## Bug description
We have seen this on our CI machines and for @sleaper even on his local machine.
Here's how it looks like with pnpm:
![image](https://user-images.githubusercontent.com/1305378/178558590-5c8a4008-400e-4d92-b4fc-116763667197.png)

here's an example output of stuck `install` with yarn:
![image](https://user-images.githubusercontent.com/1305378/178558705-63cacd8c-23d9-4430-b922-68368969d6bb.png)

My suggestions would be to remove the whole https://github.com/YassinEldeeb/prisma-tgql-types-gen/blob/26bc06ebd6540776410fd4da172c46251181ba57/packages/generator/src/generator.ts#L29

because generators often run inside a `yarn install` already.
It would be simpler and more robust to just let the use install this manually.

## To Reproduce
1. run yarn install with a schema which wants to install graphql-scalars
2. wait for the `install` to get stuck

## Expected behavior
yarn install does not get stuck

## Screenshots
:arrow_double_up: 

## Prisma information
add the used `schema.prisma` file

## Environment & setup
we have seen this on ubuntu 20 and 22 with any nodejs version.

## Prisma  Version
4.0.0

**Comment by YassinEldeeb** [comment]
Interesting! Can you provide me with further information?

> running npm/yarn install when an install is already in progress can break the in progress deps install
I can't fully understand what that means. Why would you install your dependencies while you run the `prisma generate` command? 🤔

**Comment by capaj** [comment]
@prisma/client runs prisma generate in its postinstall script. Not sure why.

**Comment by YassinEldeeb** [comment]
> @prisma/client runs prisma generate in its postinstall script. Not sure why.

Yes, It was a requested feature from the community. Cause they often forget to generate the client after installing the dependencies. So it runs the `prisma generate` command runs automatically after installing the dependencies.

**Comment by capaj** [comment]
when yarn/npm/pnpm runs postinstall script, the process is not finished yet. Lockfile is not yet written out to the file system. Running another install at this moment is running two `install` commands in paralel.
Package managers are not built to handle this usecase so they fail.

**Comment by YassinEldeeb** [comment]
> when yarn/npm/pnpm runs postinstall script, the process is not finished yet. Lockfile is not yet written out to the file system. Running another install at this moment is running two install commands in paralel.
Package managers are not built to handle this usecase so they fail or get stuck.

Oh, I got what you mean. But there's something very strange about it. As you can see here in the `installPackage` helper function, I only install `graphql-scalars` after you've changed your schema and ran `prisma generate` locally(manually). So if you run in a CI environment or even someone else got your codebase. It'll already have the package in `package.json` and It'll skip installing it.

```diff
const installPackage = (useYarn: string, pkgName: string) => {
  const packageManager = useYarn ? 'yarn add' : 'npm i'

+ const hasGraphQLScalars = fs
+   .readFileSync(path.join(process.cwd(), './package.json'), 'utf-8')
+   .includes(`"${pkgName}"`)

+ if (hasGraphQLScalars) return

  logger.info(`${GENERATOR_NAME}:Installing ${pkgName}`)
  spawn(`${packageManager} ${pkgName}`, [], {
    shell: true,
    stdio: 'inherit',
  })
}
```

**Comment by capaj** [comment]
In that case this bug is probably not that important to fix, but would be worth checking out a bit more, because it is certainly happenning and it is very maddening.

**Comment by YassinEldeeb** [comment]
> ALternative cause for this stuck generation I suspect is having npx in the generator name
That's what I've also thought of. When I was developing this generator I ran into all sorts of issues when adding the provider. It gets stuck for whatever reason. Sometimes it works with `npx`, sometimes without it. It's getting crazier and crazier. My best route to go about it now is adding the path to the bin directly like this for example `node ../node_modules/.bin/prisma-typegraphql-types-generator/bin.js`

**Comment by YassinEldeeb** [comment]
> In that case this bug is probably not that important to fix, but would be worth checking out a bit more, because it is certainly happenning and it is very maddening.

Agree. It has been bothering me for so long even in development. But I guess it's a prisma thing that we have nothing to do about it as they're saying and will be saying that this is all can fall apart in any moment.

**Comment by YassinEldeeb** [comment]
Might have to ping Prisma about it. The provider just works perfectly with `@prisma/client` but for the community generators, it's just super unreliable.
